keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,532 @@
+  - [CatchUp concurrency support](#catchup-concurrency-support)
+  - [Use master timezone instead of UTC](#use-master-timezone-instead-of-utc)
+  - [Fix applicable open issues](#fix-applicable-open-issues)
+  - [Scale Targets for GA](#scale-targets-for-ga)
+    - [CronJob Limits](#cronjob-limits)
+    - [Frequency of launched jobs](#frequency-of-launched-jobs)
+- [V1 API](#v1-api)
+- [v1beta1 changes](#v1beta1-changes)
+- [Validations](#validations)
+- [Tests](#tests)
+  - [E2E test](#e2e-test)
+    - [Existing test cases](#existing-test-cases)
+    - [New test cases](#new-test-cases)
+  - [Conformance Tests](#conformance-tests)
+  - [Unit Tests](#unit-tests)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+[CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is a Kubernetes API that creates Job object on a schedule specified by a cron spec. It is in beta status since v1.8. This document lays out the plan to promote it to stable.
+
+## Motivation
+
+CronJob definition has been stable for the last few releases and is useful to run periodic tasks in kubernetes cluster. This API adds the ability to add cron facility to a cluster. We feel the API is ready to be promoted to Stable and be supported longterm by the community.
+
+## Goals
+
+- Rearchitect CronJob controller
+- .status.lastSuccessfulTime
+- .status.nextRunTime
+- Use master timezone instead of UTC
+- Fix open issues and add tests
+
+## Non-Goals
+- Changing API field or meaning
+- [When creating a cronjob, Reference an existing job](https://github.com/kubernetes/kubernetes/issues/81329)
+- Add `CatchUp` concurrency policy
+
+## Proposal
+
+### Promote CronJob API to v1
+There is already a [batch/v1/CronJob](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/v1/register.go#L28) currently. We will formally promote the API to GA. In addition the older versions of CronJob API (batch/v1beta1, batch/v2alpha1) will be deprecated following the [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) 
+
+### Rearchitect CronJob controller
+
+#### Existing controller
+
+The current implementation of the CronJob controller is different that the other workload controllers. 
+Other controller use informers and caches to reduce the load on API server. The conjob controller does a periodic poll and sweep of all the objects and acts on them. Also CronJob controller has only one worker doing this. 
+
+1. syncs all CronJob objects [every 10 seconds](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L98). 
+2. Using pager library, gets all Pods and all CronJobs and [processes them one by one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L136)
+
+This is not a scalable design and ends up loading the API server.
```

Comment:
Can we quantify the limits of the current design so we can compare it with the performance of the new implementation?

Reply:
Hmm. We would need to add tests for this. Will do as part of development of the new controller to quantify the same.
As we write tests and make measurements we can update the KEP. For now it is given that we need an informers based rearch. Finetuning the parametrs of the new controller would be done as we test it out.

Reply:
Isn't sig-scalability capable of giving us those numbers. From what I understood from @liggitt there were some tests performed. Eventually, don't we have any issue describing just that?

Reply:
This is what i was trying to figure out. @wojtek-t  - Any inputs on how to quantify this ?

Reply:
https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds

Doesnt have any inputs for Jobs. we could extrapolate something from Pods.